GLTFLoader KHR_materials_unlit extension support.

robertlong opened the pull request:
Adds support for the pending [KHR_materials_unlit](https://github.com/KhronosGroup/glTF/pull/1163) glTF extension.

When this extension exists on a material the THREE.MeshBasicMaterial is used. The `.map` and `.color` properties are set appropriately.

I've also added an example model from [mozilla/mr-social-client](https://github.com/mozilla/mr-social-client) that shows off the extension. When selecting `glTF-MaterialsUnlit` in the dropdown a model with the extension is loaded and THREE.MeshBasicMaterial is used. When you select `glTF` a model without the extension is loaded and THREE.StandardMaterial is used with the correct fallback properties.

I could use some feedback on how to handle occlusion, emmisive, and normal maps. Exporters are expected to exclude these maps. I can ensure that these maps are removed from the material or I can leave it how it is and expect the exporters to remove them.

I also have a PR (https://github.com/facebookincubator/FBX2glTF/pull/61) to export FBX models as unlit glTF models if you wish to test the loader with your own models.

robertlong committed:
GLTFLoader KHR_materials_unlit extension support.

donmccurdy commented:
Thanks! The code looks good to me.

Here's a demo of my viewer running this version of GLTFLoader: https://khr-materials-unlit-bhyuetkweu.now.sh/

^ /cc @AurL would you want to try any of your examples to see if this looks reasonable? Here's one, edited by hand:
<img width="1440" alt="screen shot 2018-01-19 at 2 09 21 pm" src="https://user-images.githubusercontent.com/1848368/35173974-69cd83fa-fd22-11e7-9927-19d05c7663c6.png">

mrdoob commented:
Are you adding two versions (`gltf/bin` and `glb`) of the same model intentionally?

donmccurdy commented:
The `gltf/bin` model does not use the unlit extension, it is standard metal/rough for comparison. No preference on whether it should be included though.

mrdoob commented:
As far as I understand `glb` also includes the textures, so we're adding the textures twice too...

donmccurdy commented:
Yeah if keeping the repo size down is a concern, fine to remove glTF/bin version. Probably have more glTF examples than we need already. (cute robot though! 😊)

WestLangley commented:
@donmccurdy Do you have a live example showing `MeshStandardMaterial` and the unlit material side-by-side?

donmccurdy commented:
@WestLangley sure —

Robot:
- [unlit](https://khr-materials-unlit-bhyuetkweu.now.sh/#model=https://cdn.rawgit.com/robertlong/three.js/dccc4242c43d4a42b90fff7183d9873945816374/examples/models/gltf/BotSkinned/glTF-MaterialsUnlit/Bot_Skinned.glb)
- [standard](https://khr-materials-unlit-bhyuetkweu.now.sh/#model=https://cdn.rawgit.com/robertlong/three.js/dccc4242c43d4a42b90fff7183d9873945816374/examples/models/gltf/BotSkinned/glTF/Bot_Skinned.gltf)

FTM ([original](https://sketchfab.com/models/0970f30574d047b1976ba0aa6f2ef855))
- [unlit](https://khr-materials-unlit-bhyuetkweu.now.sh/#model=https://storage.googleapis.com/blocks-town/tmp/ftm_unlit/scene.gltf)
- [standard](https://khr-materials-unlit-bhyuetkweu.now.sh/#model=https://storage.googleapis.com/blocks-town/tmp/ftm/scene.gltf) (need to turn off envMap in lighting settings)

Main expected use cases here are improved performance on mobile devices for VR, and photogrammetry.

mrdoob commented:
Yes please, lets remove the standard metal/rough version from the PR.

WestLangley commented:
I must be missing something... Don't you want to approximate a PBR material with an unlit one?

If that is the case, I think you should be using `MeshBasicMaterial` with an enviornment map and dialing the `reflectivity` based on the `metalness`. You will have to select the best `combine` value, too.

donmccurdy commented:
@WestLangley that might be the goal for this particular robot model, I'm not sure... But broadly the KHR_materials_unlit extension is only to specify "the model should be unlit", and the rest is up to the user. So this could also apply to 3D scans or anime/drawing styles where we do not want to approximate PBR at all. See https://github.com/KhronosGroup/glTF/issues/1215.

takahirox commented:
Off topic tho, I often think when the best timing of supporting new glTF spec/extension for `GLTFLoader` is.

1. Made PR/issue in glTF GitHub and started discussion
2. Likely almost fixed
3. Fixed and published in https://github.com/KhronosGroup/glTF/tree/master/specification/2.0 or https://github.com/KhronosGroup/glTF/tree/master/extensions

`KHR_materials_unlit` is in the phase 2?

donmccurdy commented:
Yes, I would say `KHR_materials_unlit` is in "likely almost fixed" phase. The list of extensions in [glTF/extensions](https://github.com/KhronosGroup/glTF/tree/master/extensions) are not necessarily farther along though, e.g. KHR_technique_webgl is still a "draft" and further from finished than this one, even though this one is still a PR.

As far as when to merge, I don't have a strong preference. I think I was too early with KHR_lights in glTF 2.0 as that one is still changing, but fewer moving parts here. :)

takahirox commented:
I see the status, thanks.

How about merging for r91?

(If my guess is right) seems like r90 will be released soon, and there seems be a discussion left a bit for `KHR_materials_unlit` https://github.com/KhronosGroup/glTF/pull/1163#issuecomment-358542442

donmccurdy commented:
Sure, r91 seems fine to me.

zellski commented:
I believe if `material.emissiveFactor` or `material.emissiveTexture` are present in the glTF file, they will override `materialParams.color` and `materialParams.map` a little further down in the same function. The extension PR not explicit about what should happen if emissive properties are present (except to suggest they should not be), but it does request:
```
color = <baseColorTerm>
```
which suggests the emissive terms should be explicitly ignored in the presence of this extension. Thoughts?

donmccurdy commented:
As the extension is written, agreed that emissive should be stripped. Would also be open to removing that stipulation in the spec, since emissive isn't necessarily incompatible with an unlit material.

zellski commented:
I've come across the conflict only once, and it was clearly a mistake -- someone had left `emissiveFactor`: [0, 0, 0] in place, and they couldn't work out why everything was pitch black even though KHR_materials_unlit was in place, and the baseColor* properties were correct. While it was a little perplexing that emissive was still relevant, this is clearly an edge case.

Are there good reasons to actually include emissive elements in the fallback material? I know I should ask this in the extension PR, but I'm terrified of disrupting the fragile tranquility that we've achieved there. :-)

donmccurdy commented:
I can't think of a reason that emissive would be used in the fallback material, since it would necessarily be added to baseColor anyway... The use case I envisioned was that someone might want an emissive map on their (supported) unlit material, e.g. an unlit VR mobile scene with glowy lights. But choosing unlit for performance and then also wanting post processing might be an unlikely combination, I'm not sure...

zellski commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -397,6 +404,51 @@ THREE.GLTFLoader = ( function () {
+		materialParams.color = new THREE.Color( 1.0, 1.0, 1.0 );
+		materialParams.opacity = 1.0;
+
+		if ( Array.isArray( metallicRoughness.baseColorFactor ) ) {
```
At this point, we don't actually know that metallicRoughness is defined, do we? The `material.pbrMetallicRoughness !== undefined` test in the calling function happens after the call to this extension. I think this would NPE in a (weird but valid) material lacking that particular subobject.=

donmccurdy commented:
About emissive, a bit of old code left over from glTF 1.0:

```js
if ( materialDef.emissiveFactor !== undefined ) {

  if ( materialType === THREE.MeshBasicMaterial ) {

    materialParams.color = new THREE.Color().fromArray( materialDef.emissiveFactor );

  } else {

    materialParams.emissive = new THREE.Color().fromArray( materialDef.emissiveFactor );

  }

}
```

overwriting `materialParams.color` is not what we want, for now I think that should be:

```js
if ( materialDef.emissiveFactor !== undefined ) {

  materialParams.emissive = new THREE.Color().fromArray( materialDef.emissiveFactor );

}
```

Whether emissive should be stripped or allowed I will ask on the extension PR.

WestLangley commented:
>MeshBasicMaterial does not support emissive at all...

We could discuss adding it, but first see if you can reasonably avoid needing this feature.

donmccurdy commented:
We can avoid needing the feature.

In that case I propose changing the above code to:

```js
if ( materialDef.emissiveFactor !== undefined && materialType !== THREE.MeshBasicMaterial ) {

  materialParams.emissive = new THREE.Color().fromArray( materialDef.emissiveFactor );

}
```

robertlong committed:
Check for pbrMetallicRoughness and prevent applying additional material maps.

robertlong committed:
Remove .glb model. Update .gltf model. Update example.

robertlong commented:
@donmccurdy I made the changes you recommended and also removed setting occlusion and normal maps on the MeshBasicMaterial.

I also removed the `.glb` model. Is the example correct now? It feels a bit odd that nothing displays until you select the unlit material extension from the dropdown.

donmccurdy commented:
Thanks!

> I also removed the .glb model. Is the example correct now? It feels a bit odd that nothing displays until you select the unlit material extension from the dropdown.

Yeah, really the demo should default to the first available extension but that's a bug in our demo setup, and we can fix it later.

donmccurdy commented:
Updated demo: https://khr-materials-unlit-ooddguuati.now.sh/

Vertex colors and (lack of) emissive in Quill example look good now:

<img width="1436" alt="screen shot 2018-01-25 at 8 10 10 pm" src="https://user-images.githubusercontent.com/1848368/35424925-eb6b2b54-020b-11e8-9450-b684c0d66006.png">

zellski commented:
One tiny nit: if you are leaving `KHR_materials_common` support in for now, given that its days are probably numbered, does it make sense to check for `KHR_materials_unlit` first? It seems both the more specific request on behalf of the author, and also the more likely-to-survive-the-coming-culling way to go.

donmccurdy commented:
Will remove it shortly: https://github.com/mrdoob/three.js/pull/13169

donmccurdy commented:
I think the KHR_materials_unlit extension is stable enough to support in three.js now. Here is a change set on the Blender exporter for testing: https://github.com/KhronosGroup/glTF-Blender-Exporter/pull/160

@robertlong could you rebase this PR?

robertlong committed:
Default to first available extension.

robertlong commented:
I rebased the PR and made the glTF extensions example default to the model of the first available extension.

I think this should be good to go now.

donmccurdy commented:
Oops, I think the rebase merging has brought back the (deleted) GLTFMaterialsCommonExtension.

robertlong committed:
Remove KHR_materials_common

robertlong commented:
Oops, yeah I didn't see that the GLTFMaterialsCommonExtension made it's way back in. I've removed it. Thanks!

donmccurdy commented on examples/webgl_loader_gltf_extensions.html:
```diff
@@ -431,6 +436,14 @@
+					objectRotation: new THREE.Euler(0, 0, 0),
+					addLights:true,
+					extensions: ['glTF-MaterialsUnlit'],
+					addEnvMap: true
```
nit: No need for `addEnvMap: true` or `addLights: true` on this one.

donmccurdy commented:
Thanks! Looks good to me. 🚀

mrdoob commented:
Thanks!

takahirox commented:
IMO, the unlit example is a bit hard to see. Better to change background?

![image](https://user-images.githubusercontent.com/7637832/36950897-162bd562-2040-11e8-8427-ced65ec08a81.png)
